File: downloads/code/LeetCode/Python/Game-of-Life.py
class Solution(object):
    def countLiveNeighbors(self, i,j,board,m,n):
        dx = [-1,-1,-1,0,0,1,1,1]
        dy = [-1,0,1,-1,1,-1,0,1]
        
        cnt = 0
        # 8 个邻居
        for k in range(8):
            nx = dx[k]+i
            ny = dy[k]+j
            if nx < 0 or ny < 0 or nx >=m or ny >= n:
                continue
            if board[nx][ny] == 1 or board[nx][ny] == 2:
                cnt += 1
        return cnt
    
    def gameOfLife(self, board):
        """
        :type board: List[List[int]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        

        m = len(board)
        n = len(board[0])

        if m == 0 or n == 0:
            return

        for i in range(m):
            for j in range(n):
                cnt = self.countLiveNeighbors(i,j,board,m,n)
                # board[i][j] != 0
                if board[i][j]:
                    if cnt != 2 and cnt != 3:
                        board[i][j] = 2
                else:
                # board[i][j] == 0
                    if cnt == 3:
                        board[i][j] = 3
        for i in range(m):
            for j in range(n):
                board[i][j] &= 1

File: downloads/code/LeetCode/Java/Game-of-Life.java
public class Solution {
    public int cnt(int[][] board, int row, int col) {
        int[] dx = {0,0,-1,-1,-1,1,1,1};
        int[] dy = {1,-1,0,-1,1,-1,0,1};

        int m = board.length;
        int n = board[0].length;

        int ret = 0;
        for(int i=0;i<8;i++){
            int x = dx[i]+row;
            int y = dy[i]+col;
            if(x<0 || y<0 || x>=m || y>=n) continue;
            if(board[x][y]==1 || board[x][y]==2) ret++;

        }
        return ret;

    }
    public void gameOfLife(int[][]board){
        int m=board.length;
        int n=board[0].length;
        if(m==0 || n==0)
            return ;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                int count = cnt(board,i,j);
                if(board[i][j] != 0){
                    if(count!=2 && count!=3)
                        board[i][j] = 2;
                }
                else{
                    if(count==3)
                        board[i][j] = 3;
                
                }
            }
        }
        for(int i=0;i<m;i++)
            for(int j=0;j<n;j++)
                board[i][j] &= 1;

    }
}
